### ISSUE TITLE
DecodePcodeExecutionException occurs when using BytesTracePcodeEmulator

### ISSUE THREAD

**Issue by matthewpruett** [NationalSecurityAgency/ghidra#6017]
**Describe the bug**
Using the `BytesTracePcodeEmulator` in a simple script results in a `DecodePcodeExecutionException` on the first call to `stepInstruction()`.

**To Reproduce**
1. Open an x86-64 binary in ghidra
2. Modify the address on line 5 of the provided script to be the address of the current program's `main` function.
3. Run the script

**Expected behavior**
The script should emulate the first instruction and print the address of the next instruction.

**Attachments**
Ghidra script
```
from ghidra.app.plugin.core.debug.service.emulation import ProgramEmulationUtils
from ghidra.pcode.exec.trace import BytesTracePcodeEmulator
from ghidra.pcode.exec.PcodeExecutorStatePiece import Reason

trace = ProgramEmulationUtils.launchEmulationTrace(currentProgram, toAddr(0x10064a), 1)
traceThread = trace.getThreadManager().getAllThreads().iterator().next()
host = trace.getPlatformManager().getHostPlatform()

emulator = BytesTracePcodeEmulator(host, 0)
thread = emulator.newThread(traceThread.getPath())

rip = currentProgram.getLanguage().getRegister("RIP")
print(thread.getState().getVar(rip, Reason.INSPECT))
thread.stepInstruction()
print(thread.getState().getVar(rip, Reason.INSPECT))
```

Output
```
new_emu_trace.py> Running...
array('b', [74, 6, 16, 0, 0, 0, 0, 0])
Traceback (most recent call last):
  File "new_emu_trace.py", line 14, in <module>
    thread.stepInstruction()
	at ghidra.pcode.emu.SleighInstructionDecoder.decodeInstruction(SleighInstructionDecoder.java:91)
	at ghidra.pcode.emu.DefaultPcodeThread.executeInstruction(DefaultPcodeThread.java:582)
	at ghidra.pcode.emu.DefaultPcodeThread.stepInstruction(DefaultPcodeThread.java:417)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
ghidra.pcode.exec.DecodePcodeExecutionException: ghidra.pcode.exec.DecodePcodeExecutionException: Unknown disassembly error (PC=0010064a)
new_emu_trace.py> Finished!
```

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04.6
 - Java Version: 17.0.8.1
 - Ghidra Version: 10.4
 - Ghidra Origin: official GitHub distro

**Comment by nsadeveloper789** [comment]
**TL;DR:** You're probably looking for `BytesDebuggerPcodeEmulator`. Forgive any errors in my untested Python code below.

First correct some imports:
```python
from ghidra.app.plugin.core.debug.service.emulation.data import DefaultPcodeDebuggerAccess
from ghidra.app.plugin.core.debug.service.emulation import BytesDebuggerPcodeEmulator # Same package as ProgramEmulationUtils
```

You don't need `BytesTracePcodeEmulator`.
Then, correct how you create the emulator:
```python
emulator = BytesDebuggerPcodeEmulator(DefaultPcodeDebuggerAccess(state.tool, None, host, 0))
```

 I think everything should be good after that.

# Long version

You're mixing two different APIs for Emulation. There are actually 3 "tiers," and you're mixing the 2nd and 3rd. The 1st tier is just a plain emulator: `PcodeEmulator`. It has a state totally dissociated from anything else. It stands alone. Thus, you must explicitly move state in and out of it, including the bytes from the initial image.

The 2nd tier `BytesTracePcodeEmulator` introduces the `Trace` as a place to implicitly retrieve and store state. The emulator's internal state is just used as a cache. Reads from uninitialized parts of the state go through to the trace at the bound snapshot key (0 in your code). There's also a convenience method `TracePcodeMachine.writeDown` (inherited by the emulator) that lets you dump the emulator's current state back into the trace at a chosen snapshot.

The 3rd tier `BytesDebuggerPcodeEmulator` introduces static mappings and targets as a place to implicitly retrieve state (but not store) state. Again, the emulator's state is a cache, and reads from uninitialized parts to through to the trace. If the trace is also uninitialized, then it'll query the target, if applicable. Finally, it'll check the static mappings and try to read bytes from a mapped image (Ghidra program).

The first call you made `ProgramEmulationUtils.launchEmulationTrace` merely maps the image into the trace. It *does not* copy the image into the trace, nor into any emulator. It's just setting things up for a Tier 3 emulator.

To figure out what tier a class "belongs" to, check what Ghidra Module it is in. Tier 1 stuff is all in "Framework Emulation". Tier 2 stuff is all in "Debug Framework-TraceModeling". Tier 3 stuff is all in "Debug Debugger". The tier you select depends on your use case. For the tightest UI integration, but perhaps slowest performance and greatest number of dependencies, select Tier 3. For generating traces that you can later explore in the UI, select Tier 2. For raw emulation, having the fewest dependencies, select Tier 1.

**Comment by matthewpruett** [comment]
Thank you for the explanation of the different tiers of emulators. I tried switching to the `BytesDebuggerPcodeEmulator` using your instructions, but I'm unfortunately getting the same exception.

**Comment by nsadeveloper789** [comment]
Oof. Okay. Perhaps I'll give it a go myself sometime later. In the meantime, I have a guess:

The trace needs to be open in the tool, otherwise the static mapping service will not process it. Something like:

```python
traceManager = state.tool.getService(TraceManagerServce)
traceManager.openTrace(trace)
traceManager.activateThread(traceThread) # Optional
trace.release(1) # Allow the manager to be the sole consumer
```

Refer to `DebuggerEmuExampleScript.java`, starting around line 117.

**Comment by matthewpruett** [comment]
That worked. Thank you for the help!
